### ISSUE TITLE
.data and .detach 

### ISSUE THREAD

**Issue by easonnie** [pytorch/pytorch#7860]
I try to play the `detach` function with the code below.

```
    print(torch.__version__)

    a = torch.Tensor([1, 2])
    a.requires_grad_(True)

    b = torch.Tensor([2, 3])
    b.requires_grad_(True)

    y = torch.Tensor([1, 1])
    y.requires_grad_(True)

    x = a * b
    x_detach = x.detach()

    print("x = a * b (a = 2, b = 3)")
    print("x require gard:", x.requires_grad)
    print("x.detach() require gard:", x_detach.requires_grad)

    print('-' * 6, "Before Fill", '-' * 6)
    print("x:", x)
    print("x.detach():", x_detach)

    print('-' * 6, "After x.detach().zero_()", '-' * 6)
    x_detach.zero_()
    print("x:", x)
    print("x.detach():", x_detach)

    print("out = (x * y).sum()")
    print("x.backward()")
    out = (x * y).sum()
    out.backward()

    print("out:", out)
    print("x.grad:", x.grad)
    print("y.grad:", y.grad)
    print("a.grad:", a.grad)
    print("b.grad:", b.grad)
```

The output is:
```
0.4.0
x = a * b (a = 2, b = 3)
x require gard: True
x.detach() require gard: False
------ Before Fill ------
x: tensor([ 2.,  6.])
x.detach(): tensor([ 2.,  6.])
------ After x.detach().zero_() ------
x: tensor([ 0.,  0.])
x.detach(): tensor([ 0.,  0.])
out = (x * y).sum()
x.backward()
out: tensor(0.)
x.grad: None
y.grad: tensor([ 0.,  0.])
a.grad: tensor([ 2.,  3.])
b.grad: tensor([ 1.,  2.])
```

The results seem to contradict the description below:

`.data` was the primary way to get the underlying `Tensor` from a `Variable`. After this merge, calling `y = x.data` still has similar semantics. So `y` will be a Tensor that shares the same data with `x`, is unrelated with the computation history of `x`, and has `requires_grad=False`.

However, `.data` can be unsafe in some cases. Any changes on `x.data` wouldn't be tracked by `autograd`, and the computed gradients would be incorrect if `x` is needed in a backward pass. A safer alternative is to use `x.detach()`, which also returns a `Tensor` that shares data with `requires_grad=False`, but will have its in-place changes reported by `autograd` if `x` is needed in backward.

Shouldn't it raise an Exception?

**Comment by colesbury** [comment]
No -- the protection is only to detect modifications to (hidden) internal values saved for backwards. It's does not throw exceptions on explicit modifications to variables.

For example, this will throw an exception:

```python
x = torch.randn(1, requires_grad=True)
y = torch.randn(1, requires_grad=True)
z = x * y
x_detach = x.detach()
x_detach.zero_()
z.backward()   # throws
```

This will NOT throw an exception


```python
x = torch.randn(1, requires_grad=True)
y = torch.randn(1, requires_grad=True)
z = x * y
z_detach = z.detach()
z_detach.zero_()
z.backward()   # throws
```

In the first case, x is saved internally needed to compute the gradient of dz/dy. In the second case, modifying z directly does not cause an exception to be thrown.
